src/lib/work_partitioner/job_pool.ml

```diff
@@ -0,0 +1,125 @@
+open Core_kernel
+
+type ('accum, 'final) fold_action =
+  { action : [ `Continue of 'accum | `Stop of 'final ]; slashed : bool }
+
+let epoch_now () = Time.(now () |> to_span_since_epoch)
+
+module Make (ID : Hashtbl.Key) (Spec : T) = struct
+  type job = (Spec.t, ID.t) Snark_work_lib.With_job_meta.t
+
+  type job_item = job option ref
+
+  type t =
+    { timeline : job_item Deque.t (* For iteration  *)
+    ; index : (ID.t, job_item) Hashtbl.t (* For marking job as done *)
+    }
+
+  let create () =
+    { timeline = Deque.create (); index = Hashtbl.create (module ID) }
+
+  let rec peek (t : t) =
+    match Deque.dequeue_front t.timeline with
+    | None ->
+        None
+    | Some { contents = None } ->
```

Comment:
While deconstructing of ref with `contents` is allowed by Ocaml, this is the first time I see it in our codebase, took me a while to understand where does the `contents` come from.

I'd suggest to rewrite it in line with existing practice:
```ocaml
| Some job_item ->
    ( match !job_item with | None -> peek t | Some pending -> 
        (Deque.enqueue_front t.timeline job_item ;
        Some pending))
```

Also, I wonder, how this function is going to be used? I worry that by peeking we may create a potential for a race condition: if some code uses the `pending`, and in between `pending` to be taken and `pending` to be used there is some action that replaces `Some pending` with `None`, this might be problematic.